server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -128,27 +129,27 @@ public Releasable forceRefreshes() {
      *        false otherwise.
      * @return did we call the listener (true) or register the listener to call later (false)?
      */
-    public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listener) {
+    public boolean addOrNotify(Translog.Location location, Consumer<AsyncRefreshResult> listener) {
```

Comment:
I wonder if we really need to make this change and the segment generation supplier at all?

My understanding (please correct me if I'm wrong) is that we don't need the refreshed segment generation on stateful Elasticsearch and we need it only for index shards on stateless. And in that case it is already provided through `InternalEngine#refresh(String)` when the write request's refresh policy is `IMMEDIATE` and when it is `WAIT_UNTIL` we only need to wait for the next refresh and then retrieve the latest commited segment generation which is already available with `getLastCommittedSegmentInfos`?

Happy to discuss this

Reply:
This seems like a good idea. I have access to the `IndexShard` in post refresh action. And I can get that at minimum from the commit stats. I'll make this change tomorrow.

Reply:
I mean, I guess the biggest concern here is that this requires a flush happens for this refresh logic to work. Obviously this occurs in stateless, but all of the "unpromotable" logic really doesn't involve stateless at all. So I guess I would need a discoveryNode.isStateless check here.

Reply:
We should maybe talk about this briefly tomorrow. Currently the refresh listener is called from the Lucene level. So I think that we would need to change the IndexEngine logic to ensure that a refresh call is ALWAYS forcing a flush. Right now we only do that from the api call. So without a change there, the periodic refresh could indicate that the doc is visible, but the generation still be the old generation (I think).